### PR TITLE
`chain-spec`/presets reference docs added

### PR DESCRIPTION
**Pull request by michalkucharczyk**
Added reference doc about:
- the pallet genesis config and genesis build, 
- runtime `genesis-builder` API,
- presets,
- interacting with the `chain-spec-builder` tool

I've added [minimal runtime](https://github.com/paritytech/polkadot-sdk/tree/mku-chain-spec-guide/docs/sdk/src/reference_docs/chain_spec_runtime) to demonstrate above topics.

I also sneaked in some little improvement to `chain-spec-builder` which allows to parse output of the `list-presets` command.

### CODE REVIEW

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 1]
```diff
@@ -1,4 +1,107 @@
...
+//! - network / logical properties of the chain, the most important property being the list of
+//!   bootnodes.
+//!
+//! This document describes how initial state is handled in pallets and runtime, and how to interact
+//! with runtime in order to build genesis sate.
+//!
+//! For more informations on chain specification and its properties refer to
+//! [`sc_chain_spec#from-initial-state-to-raw-genesis`].
+//!
+//! The intial genesis state can be prvovided in the following formats:
```
Comment (lexnv) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:17]:
```suggestion
//! The initial genesis state can be provided in the following formats:
```

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 2]
```diff
@@ -1,4 +1,107 @@
...
+//!   - full,
+//!   - patch,
+//!   - raw,
+//!
+//! Each of the format is explained in [_chain-spec-format_][`sc_chain_spec#chain-spec-formats`].
+//! The
```
Comment (lexnv) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:23]:
```suggestion
//!
```

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 3]
```diff
@@ -1,4 +1,107 @@
...
+//! The initial state conveyed in `GenesisConfig` struct is transformed into state storage items by
+//! means of the [`BuildGenesisConfig`] trait wich shall be implemented for pallet's `GenesisConfig`
+//! struct. The [`pallet::genesis_build`] attribute shall be attached to the `impl` block:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_bar_build)]
+
+//! GenesisConfig may contain also more complicated types including nested structs or enums, as in
+//! the example for `pallet_foo`:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_GenesisConfig)]
+
+//! Please not that fields of `GenesisConfig` may not be directly mapped to storage items:
```
Comment (lexnv) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:43]:
```suggestion
//! Please note that fields of `GenesisConfig` may not be directly mapped to storage items:
```

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 4]
```diff
@@ -1,4 +1,107 @@
...
+//! [`build_state`], [`get_preset`].
+//! Usual implementation of [`sp_genesis_builder::GenesisBuilder`] is as follows:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/runtime.rs", runtime_impl)]
+
+//! Please note that two functions are customized `preset_names` and `get_preset`. The first one
+//! just provdes a vec of names of supported presets, while the latter one delegats the call to
+//! function that maps the name to preset: [`chain_spec_guide_runtime::presets::get_builtin_preset`]
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/presets.rs", get_builtin_preset)]
+
+//! ## Genesis state presets for `runtime`
```
Comment (lexnv) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:64]: nit: I think we can leave a blank line after this for consistency with above

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 5]
```diff
@@ -1,4 +1,110 @@
...
+//! GenesisConfig may contain also more complicated types including nested structs or enums, as in
+//! the example for `pallet_foo`:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_GenesisConfig)]
+
+//! Please note that fields of `GenesisConfig` may not be directly mapped to storage items:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_build)]
+
+//! ## `GenesisConfig` for `runtimes`
+
+//! The runtime genesis config struct consists of configs for every pallet. For [_demonstration
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:48]: There's good explanation of this also in the `runtime_types` ref doc, maybe you can reuse some of that.

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 6]
```diff
@@ -1,4 +1,110 @@
...
+//! means of the [`BuildGenesisConfig`] trait wich shall be implemented for pallet's `GenesisConfig`
+//! struct. The [`pallet::genesis_build`] attribute shall be attached to the `impl` block:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_bar_build)]
+
+//! GenesisConfig may contain also more complicated types including nested structs or enums, as in
+//! the example for `pallet_foo`:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_GenesisConfig)]
+
+//! Please note that fields of `GenesisConfig` may not be directly mapped to storage items:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_build)]
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:44]: I didn't understand where the snippt of `pallet_foo_build` demonstrates this point, but the point is valid.

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 5 above)
Comment (kianenigma) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:48]: or at least link to as a further reading link.

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 7]
```diff
@@ -1,4 +1,110 @@
...
+
+//! Internally presets can be provided in a number of ways:
+//! - json in string form:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/presets.rs", preset_1)]
+//! - json using runtime types to serialize values:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/presets.rs", preset_2)]
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/presets.rs", preset_3)]
+//! It is worth noting that preset does not have to be the full `RuntimeGenesisConfig`, in that
+//! sense that it does not have to contain all the keys of the struct. The preset is acutally a json
+//! patch that will be merged with default value of `RuntimeGenesisConfig`. This approach should
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:80]: to be nit picky I would say the concept of "Default" for genesis config is not explained up until this point. 

Worth saying:

- `GenesisConfig` of all pallets _must_ have a `Defualt`
- which is aggregated into `RuntimeGenesisConfig`'s `Default`.
- this defined what the `get_preset` runtime api would return with `id: None`

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 8]
```diff
@@ -1,4 +1,110 @@
...
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/presets.rs", preset_4)]
+
+//! # Interacting with the tool
+//!
+//! The [`chain_spec_builder`] util allows to interact with runtime in order to list or display
+//! presets and build the chain specification file. It is possible to use the tool with the
+//! [_demonstration runtime_][`chain_spec_guide_runtime`]. To build required packages just run the
+//! following command:
+//! ```ignore
+//! cargo build -p staging-chain-spec-builder -p chain-spec-guide-runtime --release
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:92]: or as of recently we can do `cargo install staging-chain-spec-builder` right?

File: docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
+use serde_json::{json, Value};
+use std::{process::Command, str};
+
+const WASM_FILE_PATH: &str =
+	"../../../../../target/release/wbuild/chain-spec-guide-runtime/chain_spec_guide_runtime.wasm";
+
+const CHAIN_SPEC_BUILDER_PATH: &str = "../../../../../target/release/chain-spec-builder";
+
+fn get_chain_spec_builder_path() -> &'static str {
+	// dev-dependencies do not build binary. So let's do the naive work-around here:
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs:10]: I have been brainstorming with how e can format this nicely into the docs with @rzadp, and the outcome is https://github.com/sam0x17/docify/issues/26

File: docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+				}
+			},
+			"someInteger": 100
+		},
+	});
+	assert_eq!(output, expected_output, "Output did not match expected");
+}
+
+#[test]
+#[docify::export]
```
Comment (kianenigma) [inline docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs:78]: In general, I think some of the docify export snippets are too long. For example, for this one, should not just export the `Command`? the rest of it is just for our testing.

**Review by kianenigma — APPROVED** [review]
All in all, this is a superb documentation, very aligned with how it should be.

I hope you can testify that writing this is mostly _writing new e2e tests_ that in general ensure the code you are writing is correct. The doc part is a small part of it that we extract. 

I hope to see more important APIs to be written like this from day-zero, with documentation in mind. A lot of the e2e testing and then actually live inside the `polkadot-sdk-docs` crate. 

Some nitpicks pointed out, but all in all looks great. 

This motivates me also bring https://github.com/paritytech/polkadot-sdk/pull/3946 up to date and merge it.

File: docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs (hunk 10 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_runtime/tests/chain_spec_builder_tests.rs:78]: My intention here was also to show the output of the command. I think it has some educative value.

Ideally if we could have macro like this:
```
command_line_sample!({
  ls -1
}, {
  output_file1,
  output file2 
})
```

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 7 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:80]: good point, thx

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 6 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:44]: ok, I've added sentence to explain it better.

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs [hunk 11]
```diff
@@ -1,4 +1,110 @@
...
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_GenesisConfig)]
+
+//! Please note that fields of `GenesisConfig` may not be directly mapped to storage items:
+#![doc = docify::embed!("./src/reference_docs/chain_spec_runtime/src/pallets.rs", pallet_foo_build)]
+
+//! ## `GenesisConfig` for `runtimes`
+
+//! The runtime genesis config struct consists of configs for every pallet. For [_demonstration
+//! runtime_][`chain_spec_guide_runtime`] used in this guide, it consists of `SystemConfig`,
+//! `BarConfig` and `FooConfig`. This structure was automatically generated by macro and it can be
```
Comment (skunert) [inline docs/sdk/src/reference_docs/chain_spec_genesis.rs:50]: I can not find `SystemConfig`, `BarConfig` and `FooConfig` anywhere, maybe leftovers from an older version of this doc?

Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:50]: Macro generated names, just click the link provided for sneak peaking...

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 6 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:51]: see: c5d1f865eb4

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 7 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:104]: added _default_ section in: c5d1f865eb4

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 8 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:116]: added note in: bf15ed3

File: docs/sdk/src/reference_docs/chain_spec_genesis.rs (hunk 5 above)
Comment (michalkucharczyk) [reply docs/sdk/src/reference_docs/chain_spec_genesis.rs:55]: added link: c5d1f865eb4

**Review by skunert — APPROVED** [review]
Nice!

**Pull request merged**